File: 그래프_서로소집합(union-find)/leetcode1319(Medium)-Number of Operations to Make Network Connected.js
// 유니온파인드 풀이 : 102ms(94.74%), 55.2MB(89.47%)

const findParent = (parent, x) => {
  if (parent[x] !== x) {
    parent[x] = findParent(parent, parent[x])
  }
  return parent[x]
}

const unionParent = (parent, a, b) => {
  const x = findParent(parent, a)
  const y = findParent(parent, b)
  if (x < y) {
    parent[y] = x
  } else {
    parent[x] = y
  }
}

/**
 * @param {number} n
 * @param {number[][]} connections
 * @return {number}
 */
var makeConnected = function (n, connections) {
  if (connections.length < n - 1) {
    return -1
  }

  const parent = Array.from({ length: n }, (v, i) => i)
  let edges = n - 1
  for (const [v, w] of connections) {
    if (edges === 0) {
      return 0
    }
    if (findParent(parent, v) !== findParent(parent, w)) {
      edges -= 1
      unionParent(parent, v, w)
    }
  }

  return edges
}

File: 그래프_유니온파인드(최소신장트리, 크루스칼)/leetcode1319(Medium)-Number of Operations to Make Network Connected.py
# 유니온파인드 풀이 : 465ms(97.9%), 34MB(76.41%)

def find_parent(parent, x):
    if parent[x] != x:
        parent[x] = find_parent(parent, parent[x])
    return parent[x]

def union_parent(parent, a, b):
    x, y = find_parent(parent, a), find_parent(parent, b)
    if x < y:
        parent[y] = x
    else:
        parent[x] = y

class Solution:
    def makeConnected(self, n: int, connections: List[List[int]]) -> int:
        if len(connections) < n - 1:
            return -1

        parent = [x for x in range(n)]
        edges = n - 1
        for v, w in connections:
            if edges == 0:
                return 0
            if find_parent(parent, v) != find_parent(parent, w):
                union_parent(parent, v, w)
                edges -= 1

        return edges